--- a/backend/drm/backend.c	2023-02-10 19:29:59.000000000 -0300
+++ b/backend/drm/backend.c	2024-04-19 00:44:58.308346962 -0300
@@ -251,10 +251,18 @@
 		goto error_event;
 	}
 
+	drm->session_destroy.notify = handle_session_destroy;
+	wl_signal_add(&session->events.destroy, &drm->session_destroy);
+
+	drm->display_destroy.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &drm->display_destroy);
+
 	if (drm->parent) {
 		if (!init_drm_renderer(drm, &drm->mgpu_renderer)) {
 			wlr_log(WLR_ERROR, "Failed to initialize renderer");
-			goto error_resources;
+			wl_list_remove(&drm->parent_destroy.link);
+			drm->parent = NULL;
+			return &drm->backend;
 		}
 
 		// We'll perform a multi-GPU copy for all submitted buffers, we need
@@ -264,7 +272,10 @@
 			wlr_renderer_get_dmabuf_texture_formats(renderer);
 		if (texture_formats == NULL) {
 			wlr_log(WLR_ERROR, "Failed to query renderer texture formats");
-			goto error_mgpu_renderer;
+			finish_drm_renderer(&drm->mgpu_renderer);
+			wl_list_remove(&drm->parent_destroy.link);
+			drm->parent = NULL;
+			return &drm->backend;
 		}
 
 		// Forbid implicit modifiers, because their meaning changes from one
@@ -281,18 +292,8 @@
 		}
 	}
 
-	drm->session_destroy.notify = handle_session_destroy;
-	wl_signal_add(&session->events.destroy, &drm->session_destroy);
-
-	drm->display_destroy.notify = handle_display_destroy;
-	wl_display_add_destroy_listener(display, &drm->display_destroy);
-
 	return &drm->backend;
 
-error_mgpu_renderer:
-	finish_drm_renderer(&drm->mgpu_renderer);
-error_resources:
-	finish_drm_resources(drm);
 error_event:
 	wl_list_remove(&drm->session_active.link);
 	wl_event_source_remove(drm->drm_event);
